command/ssh.go

```diff
@@ -127,118 +147,309 @@ func (c *SSHCommand) Run(args []string) int {
 		// be used by the user (ACL enforcement), then user should see an
 		// error message accordingly.
 		c.Ui.Output(fmt.Sprintf("Vault SSH: Role: %q", role))
+		c.role = role
 	}
 
-	data := map[string]interface{}{
-		"username": username,
-		"ip":       ip.String(),
+	// If no mode was given, perform the old-school lookup. Keep this now for
+	// backwards-compatability, but print a warning.
+	//
+	// TODO: remove in 0.9.0, convert to validation error
+	if c.mode == "" {
+		c.Ui.Warn("" +
+			"WARNING: No -mode specified. Use -mode to tell Vault which ssh\n" +
+			"authentication mode to use. In the future, you will need to tell\n" +
+			"Vault which mode to use. For now, Vault will attempt to guess based\n" +
+			"on the API response.")
+		_, cred, err := c.generateCredential()
```

Comment:
I can't think of an alternative either, but until 0.9.0, for non-CA modes, this command will create 2 sets of credentials every time. Can we _somehow_ mention this fact in the warning above so they know that adding a `-mode` will avoid that?